substrate/client/network/src/litep2p/discovery.rs

```diff
@@ -0,0 +1,523 @@
+		};
+
+		let (kademlia_config, kademlia_handle) = {
+			let protocol_names = vec![
+				kademlia_protocol_name(genesis_hash.clone(), fork_id),
+				legacy_kademlia_protocol_name(protocol_id),
+			];
+
+			KademliaConfigBuilder::new()
+				.with_known_peers(known_peers)
+				.with_protocol_names(protocol_names)
+				.build()
+		};
+
+		(
+			Self {
+				ping_event_stream,
+				identify_event_stream,
+				mdns_event_stream,
+				kademlia_handle,
+				_peerstore_handle,
+				listen_addresses,
+				find_node_query_id: None,
+				pending_events: VecDeque::new(),
+				duration_to_next_find_query: Duration::from_secs(1),
+				address_confirmations: LruMap::new(ByLength::new(8)),
+				allow_non_global_addresses: config.allow_non_globals_in_dht,
+				public_addresses: config.public_addresses.iter().cloned().collect(),
+				next_kad_query: Some(Delay::new(KADEMLIA_QUERY_INTERVAL)),
+				local_protocols: HashSet::from_iter([
+					kademlia_protocol_name(genesis_hash, fork_id),
+					legacy_kademlia_protocol_name(protocol_id),
+				]),
+			},
+			ping_config,
+			identify_config,
+			kademlia_config,
+			mdns_config,
+		)
+	}
+
+	/// Add known peer to `Kademlia`.
+	#[allow(unused)]
+	pub async fn add_known_peer(&mut self, peer: PeerId, addresses: Vec<Multiaddr>) {
+		self.kademlia_handle.add_known_peer(peer, addresses).await;
+	}
+
+	/// Add self-reported addresses to routing table if `peer` supports
+	/// at least one of the locally supported DHT protocol.
+	pub async fn add_self_reported_address(
+		&mut self,
+		peer: PeerId,
+		supported_protocols: HashSet<ProtocolName>,
+		addresses: Vec<Multiaddr>,
+	) {
+		if self.local_protocols.is_disjoint(&supported_protocols) {
+			return
+		}
+
+		let addresses = addresses
+			.into_iter()
+			.filter_map(|address| {
+				if !self.allow_non_global_addresses && !Discovery::can_add_to_dht(&address) {
+					log::trace!(
+						target: LOG_TARGET,
+						"ignoring self-reported non-global address {address} from {peer}."
+					);
+
+					return None
+				}
+
+				Some(address)
+			})
+			.collect();
+
+		self.kademlia_handle.add_known_peer(peer, addresses).await;
+	}
+
+	/// Start Kademlia `GET_VALUE` query for `key`.
+	pub async fn get_value(&mut self, key: KademliaKey) -> QueryId {
+		self.kademlia_handle
+			.get_record(RecordKey::new(&key.to_vec()), Quorum::One)
+			.await
+	}
+
+	/// Publish value on the DHT using Kademlia `PUT_VALUE`.
+	pub async fn put_value(&mut self, key: KademliaKey, value: Vec<u8>) -> QueryId {
+		self.kademlia_handle
+			.put_record(Record::new(RecordKey::new(&key.to_vec()), value))
+			.await
+	}
+
+	/// Check if the observed address is a known address.
+	fn is_known_address(known: &Multiaddr, observed: &Multiaddr) -> bool {
+		let mut known = known.iter();
+		let mut observed = observed.iter();
+
+		loop {
+			match (known.next(), observed.next()) {
+				(None, None) => return true,
+				(None, Some(Protocol::P2p(_))) => return true,
+				(Some(Protocol::P2p(_)), None) => return true,
```

Comment:
These `/p2p/` suffixes are really annoying to deal with because it's not really enforced anywhere whether it's part of the address or not. I decided to drink the kool-aid and include them in all addresses it exposes to users. Remote peer might not, however, include the `P2p` suffix in the observed address. Our listen address could be `/ip6/::1/tcp/8888/p2p/<peer id>` whereas the remote node reports our observed as `/ip6/::1/tcp/8888/` which are "the same address". Technically one of the branches could be removed (`None`, `Some(Protocol::P2p(_))`) since the `PeerId` should always be in our addresses but I would have to check if a user has any control over that address, in which case they could omit the suffix and a valid address would fail to pass this test.